x-pack/plugins/security_solution/server/lib/detection_engine/signals/threshold/find_threshold_signals.ts

```diff
@@ -73,75 +103,139 @@ export const findThresholdSignals = async ({
-    (thresholdFieldCount === 0 || i === thresholdFieldCount - 1) && threshold.cardinality?.length
-      ? { order: { cardinality_count: 'desc' } }
-      : {};
-
-  // Generate a nested terms aggregation for each threshold grouping field provided, appending leaf
-  // aggregations to 1) filter out buckets that don't meet the cardinality threshold, if provided, and
-  // 2) return the latest hit for each bucket so that we can persist the timestamp of the event in the
-  // `original_time` of the signal. This will be used for dupe mitigation purposes by the detection
-  // engine.
-  const aggregations = thresholdFields.length
-    ? thresholdFields.reduce((acc, field, i) => {
-        const aggPath = [...Array(i + 1).keys()]
-          .map((j) => {
-            return `['threshold_${j}:${thresholdFields[j]}']`;
-          })
-          .join(`['aggs']`);
-        set(acc, aggPath, {
-          terms: {
-            field,
-            ...orderByCardinality(i),
-            min_doc_count: threshold.value, // not needed on parent agg, but can help narrow down result set
-            size: 10000, // max 10k buckets
-          },
-        });
-        if (i === (thresholdFields.length ?? 0) - 1) {
-          set(acc, `${aggPath}['aggs']`, leafAggs);
-        }
-        return acc;
-      }, {})
-    : {
-        // No threshold grouping fields provided
-        threshold_0: {
-          terms: {
-            script: {
-              source: '""', // Group everything in the same bucket
-              lang: 'painless',
+  let sortKeys;
+  let result = createSearchAfterReturnType<Record<AggregateName, ThresholdAggregate>>();
+  let mergedSearchResults = createSearchResultReturnType<ThresholdAggregateContainer>();
+
+  do {
+    // Generate a composite aggregation considering each threshold grouping field provided, appending leaf
+    // aggregations to 1) filter out buckets that don't meet the cardinality threshold, if provided, and
+    // 2) return the first and last hit for each bucket in order to eliminate dupes and reconstruct an accurate
+    // timeline.
+    const aggregations: ThresholdAggregationContainer = {
+      thresholdTerms: hasThresholdFields(threshold)
+        ? {
+            composite: {
+              sources: threshold.field.map((term, i) => ({
+                [term]: {
+                  terms: {
+                    field: term,
+                  },
+                },
+              })),
+              after: sortKeys,
+              size: 10000,
+            },
+            aggs: {
+              ...leafAggs,
+              count_check: {
+                bucket_selector: {
+                  buckets_path: {
+                    docCount: '_count',
+                  },
+                  script: `params.docCount >= ${threshold.value}`,
+                },
+              },
             },
-            ...orderByCardinality(),
-            min_doc_count: threshold.value,
+          }
+        : {
+            terms: {
```

Comment:
I think the control flow for the `!hasThresholdFields` case diverges enough that it'd be worth pulling it completely out of the `do...while` loop and handling it as a separate case. It could be worth having separate functions to build the threshold alerts from the 2 types of aggregation results as well in order to avoid repeatedly checking `isCompositeBucket` in the alert building logic.

If we have separate flows, then it should be easier to remove the extra `terms` agg here. It seems like the terms agg is mostly needed to try to fit the 2 cases into the same bucketed response format, but they still don't really fit together imo. With separate functions to process the responses, we wouldn't need the "no group by fields" case to have the bucket at all. The loop becomes simpler as well.